kokoro/presubmit.sh

```diff
@@ -17,21 +19,49 @@ export XDG_RUNTIME_DIR=/var/run
 cp ./target/nextest/default/*.xml "$KOKORO_ARTIFACTS_DIR/test_logs/"
 
-mkdir -p "$KOKORO_ARTIFACTS_DIR/binaries/"
-
-# Store the git commit hash in a file.
-echo "${KOKORO_GIT_COMMIT_oak:?}" > "$KOKORO_ARTIFACTS_DIR/binaries/git_commit"
+readonly BINARY_FILE_NAME=binary
+readonly binaries_dir="${KOKORO_ARTIFACTS_DIR}/binaries"
+readonly commit_hash_dir="${KOKORO_ARTIFACTS_DIR}/git_commit"
+mkdir -p "${binaries_dir}" "${commit_hash_dir}"
 
-# Copy the generated binaries to Placer.
-export GENERATED_BINARIES=(
-    ./oak_restricted_kernel_bin/target/x86_64-unknown-none/release/oak_restricted_kernel_bin
-    ./stage0_bin/target/x86_64-unknown-none/release/stage0_bin
-    ./enclave_apps/target/x86_64-unknown-none/release/oak_echo_enclave_app
-    ./enclave_apps/target/x86_64-unknown-none/release/oak_echo_raw_enclave_app
-    ./enclave_apps/target/x86_64-unknown-none/release/oak_functions_enclave_app
-    ./enclave_apps/target/x86_64-unknown-none/release/oak_tensorflow_enclave_app
-    ./enclave_apps/target/x86_64-unknown-none/release/quirk_echo_enclave_app
+# Maps binary name to path of build artifact.
+declare -A binaries=(
+  [oak_restricted_kernel_bin]=./oak_restricted_kernel_bin/target/x86_64-unknown-none/release/oak_restricted_kernel_bin
+  [oak_echo_enclave_app]=./enclave_apps/target/x86_64-unknown-none/release/oak_echo_enclave_app
+  [oak_echo_raw_enclave_app]=./enclave_apps/target/x86_64-unknown-none/release/oak_echo_raw_enclave_app
+  [oak_functions_enclave_app]=./enclave_apps/target/x86_64-unknown-none/release/oak_functions_enclave_app
+  [oak_tensorflow_enclave_app]=./enclave_apps/target/x86_64-unknown-none/release/oak_tensorflow_enclave_app
+  [quirk_echo_enclave_app]=./enclave_apps/target/x86_64-unknown-none/release/quirk_echo_enclave_app
+  [stage0_bin]=./stage0_bin/target/x86_64-unknown-none/release/stage0_bin
 )
-cp "${GENERATED_BINARIES[@]}" "$KOKORO_ARTIFACTS_DIR/binaries/"
 
-ls -alsR "$KOKORO_ARTIFACTS_DIR/binaries"
+# Write the commit hash under the commit hash directory, with no contents,
+# so no need to open the file, just listing it is sufficient. This is only
+# for better performance.
+readonly commit_hash="${KOKORO_GIT_COMMIT_oak:?}"
+touch "${commit_hash_dir}/${commit_hash}"
+
+# Generate the importer output structure on Placer.
+for name in "${!binaries[@]}"; do
+  mkdir -p "${binaries_dir}/${name}"
+  cp "${binaries[${name}]}" "${binaries_dir}/${name}/${BINARY_FILE_NAME}"
+
+  binary_hash=$(sha256sum --binary "${binaries[${name}]}" | cut -d " " -f 1)
+  seconds_since_epoch_utc=$(stat -c "%W" "${binaries[${name}]}")
+  cat > "${binaries_dir}/${name}/binary_metadata.textproto" << __EOF__
+# proto-file: knowledge/cerebra/oak/release/binary_metadata.proto
+# proto-message: oak.BinaryMetadata
+name: "${name}"
+digests {
+  data {
+    key: 18
+    value: "${binary_hash}"
+  }
+}
+binary_path: "${BINARY_FILE_NAME}"
+commit_hash_sha1: "${commit_hash}"
+created { seconds: ${seconds_since_epoch_utc} }
```

Comment:
This looks potentially problematic due to just using string interpolation to construct what is supposed to be a text proto object. Also the rest of the script is nontrivial, and IMO almost impossible to maintain or test. Could we use a python script or something else? (maybe at some point in the future, if not now)
Keep in mind we need to think about how to make this available to other teams too. Not sure what the best way for that would be (it can be copy pasted in the short term of course, but I'd like to have a plan to improve on that in the future).

Effectively, the only thing that belongs to the repo is the list of binaries (and the build / test commands of course). Everything else could / should be factored out, possibly even to some script that lives in google3 and is just invoked by kokoro. And in this repo we only have a TOML file or similar, that lists the binaries to be processed.

Also this proto is suspiciously similar to the provenance that kokoro itself generates. I suspect the differences should be either resolved or explicitly justified.

Reply:
Agree, the freestyle textproto creation sucks. I tested it locally on my workstation, but in general it is untestable. Should we move the proto definition to the outside world? I am unaware of the possible pitfalls.

But does this show that it is not the way to go? After all we already _have_ a python in google3 that does this, it is //k/c/o/release/import_premerge.py. This was thought as a shortcut. Without the shortcut the proto definition would always stay inside google3.

Reply:
But we don't currently trigger that python script from kokoro. If we could do that, after invoking this script, but before kokoro picks up the artifacts, then I think it would work, right?

Reply:
I am going to drop this PR. Parts of it may come in handy if we want binary files which are not just named "binary", but this is currently just nice-to-have, not required.